File: api/v1/schemas/watchlist.py
# -*- coding: utf-8 -*-
"""User watchlist schemas."""

from __future__ import annotations

import re
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator
from src.utils.symbol_normalization import canonical_stock_code


class WatchlistScoreStatusContextResponse(BaseModel):
    scope: str
    fresh_means: str
    source_freshness_implied: bool
    source_authority_implied: bool


class WatchlistItemCreateRequest(BaseModel):
    symbol: str = Field(..., min_length=1, max_length=16)
    market: Literal["cn", "hk", "us"] = "cn"
    name: Optional[str] = Field(None, max_length=128)
    source: Literal["scanner"] = "scanner"
    scanner_run_id: Optional[int] = None
    scanner_rank: Optional[int] = Field(None, ge=1)
    scanner_score: Optional[float] = Field(None, ge=0, le=100)
    theme_id: Optional[str] = Field(None, max_length=64)
    universe_type: Optional[str] = Field(None, max_length=32)
    notes: Optional[str] = Field(None, max_length=500)

    @field_validator("symbol")
    @classmethod
    def _normalize_symbol(cls, value: str) -> str:
        normalized = canonical_stock_code(value).strip().upper()
        if not normalized:
            raise ValueError("symbol is required")
        if len(normalized) > 16:
            raise ValueError("symbol must be at most 16 characters")
        if not re.fullmatch(r"[A-Z0-9][A-Z0-9.\-]*", normalized):
            raise ValueError("symbol contains invalid characters")
        return normalized

    @field_validator("name", "theme_id", "universe_type", "notes")
    @classmethod
    def _normalize_optional_text(cls, value: Optional[str]) -> Optional[str]:
        normalized = str(value or "").strip()
        return normalized or None

    @field_validator("market")
    @classmethod
    def _normalize_market(cls, value: str) -> str:
        normalized = str(value or "").strip().lower()
        if normalized not in {"cn", "hk", "us"}:
            raise ValueError("market must be one of: cn, hk, us")
        return normalized


class WatchlistItemResponse(BaseModel):
    id: int
    symbol: str
    market: str
    name: Optional[str] = None
    source: str
    scanner_run_id: Optional[int] = None
    scanner_rank: Optional[int] = None
    scanner_score: Optional[float] = None
    last_scored_at: Optional[str] = None
    score_source: Optional[str] = None
    score_profile: Optional[str] = None
    score_reason: Optional[str] = None
    score_status: Optional[str] = None
    score_status_context: Optional[WatchlistScoreStatusContextResponse] = None
    score_error: Optional[str] = None
    theme_id: Optional[str] = None
    universe_type: Optional[str] = None
    notes: Optional[str] = None
    created_at: Optional[str] = None
    updated_at: Optional[str] = None
    intelligence: Optional["WatchlistIntelligenceResponse"] = None


class WatchlistOhlcvProvenanceResponse(BaseModel):
    source: str
    source_type: str
    source_label: str


class WatchlistReasonFamilyResponse(BaseModel):
    raw_code: str
    family: str
    scope: Optional[str] = None


class WatchlistScannerSourceConfidenceReasonFamiliesResponse(BaseModel):
    degradation_reason: Optional[WatchlistReasonFamilyResponse] = None
    cap_reason: Optional[WatchlistReasonFamilyResponse] = None


class WatchlistScannerReasonFamiliesResponse(BaseModel):
    cap_reason: Optional[WatchlistReasonFamilyResponse] = None
    degradation_reason: Optional[WatchlistReasonFamilyResponse] = None
    source_confidence: Optional["WatchlistScannerSourceConfidenceReasonFamiliesResponse"] = None


class WatchlistScannerIntelligenceResponse(BaseModel):
    last_score: Optional[float] = None
    last_rank: Optional[int] = None
    status: str = "unknown"
    theme: Optional[str] = None
    theme_label: Optional[str] = None
    profile: Optional[str] = None
    reason: Optional[str] = None
    last_scanned_at: Optional[str] = None
    ohlcv_provenance: Optional[WatchlistOhlcvProvenanceResponse] = None
    score_confidence: Optional[float] = None
    score_grade_allowed: Optional[bool] = None
    cap_reason: Optional[str] = None
    degradation_reason: Optional[str] = None
    source_confidence: Optional["WatchlistScannerSourceConfidenceResponse"] = None
    reason_families: Optional["WatchlistScannerReasonFamiliesResponse"] = None
    investor_signal: Optional[Dict[str, Any]] = None


class WatchlistScannerSourceConfidenceResponse(BaseModel):
    source: Optional[str] = None
    source_label: Optional[str] = None
    source_type: Optional[str] = None
    as_of: Optional[str] = None
    freshness: Optional[str] = None
    is_fallback: Optional[bool] = None
    is_stale: Optional[bool] = None
    is_partial: Optional[bool] = None
    is_synthetic: Optional[bool] = None
    is_unavailable: Optional[bool] = None
    coverage: Optional[float] = None
    score_contribution_allowed: Optional[bool] = None
    source_authority_allowed: Optional[bool] = None
    observation_only: Optional[bool] = None
    degradation_reason: Optional[str] = None
    cap_reason: Optional[str] = None


class WatchlistStrategySimulationIntelligenceResponse(BaseModel):
    lookback_days: Optional[int] = None
    forward_days: Optional[int] = None
    avg_forward_return_pct: Optional[float] = None
    hit_rate: Optional[float] = None
    avg_excess_return_pct: Optional[float] = None
    selection_count: Optional[int] = None
    data_coverage: Optional[float] = None
    status: str = "unknown"


class WatchlistBacktestIntelligenceResponse(BaseModel):
    last_result_id: Optional[int] = None
    total_return_pct: Optional[float] = None
    max_drawdown_pct: Optional[float] = None
    sharpe: Optional[float] = None
    trade_count: Optional[int] = None
    tested_at: Optional[str] = None


class WatchlistIntelligenceResponse(BaseModel):
    scanner: WatchlistScannerIntelligenceResponse = Field(default_factory=WatchlistScannerIntelligenceResponse)
    strategy_simulation: WatchlistStrategySimulationIntelligenceResponse = Field(default_factory=WatchlistStrategySimulationIntelligenceResponse)
    backtest: WatchlistBacktestIntelligenceResponse = Field(default_factory=WatchlistBacktestIntelligenceResponse)
    catalyst_exposures: Optional[List[Dict[str, Any]]] = None


class WatchlistItemListResponse(BaseModel):
    items: List[WatchlistItemResponse] = Field(default_factory=list)


class WatchlistDeleteResponse(BaseModel):
    deleted: int


class WatchlistScoreRefreshRequest(BaseModel):
    market: Optional[Literal["cn", "hk", "us"]] = None
    source: Optional[Literal["scanner"]] = None
    theme: Optional[str] = Field(None, max_length=64)
    symbols: Optional[List[str]] = None
    force: bool = False

    @field_validator("symbols")
    @classmethod
    def _normalize_symbols(cls, value: Optional[List[str]]) -> Optional[List[str]]:
        if value is None:
            return None
        normalized: List[str] = []
        seen = set()
        for item in value:
            symbol = canonical_stock_code(item).strip().upper()
            if not symbol:
                continue
            if len(symbol) > 16:
                raise ValueError("symbol must be at most 16 characters")
            if not re.fullmatch(r"[A-Z0-9][A-Z0-9.\-]*", symbol):
                raise ValueError("symbol contains invalid characters")
            if symbol not in seen:
                seen.add(symbol)
                normalized.append(symbol)
        return normalized

    @field_validator("theme")
    @classmethod
    def _normalize_theme(cls, value: Optional[str]) -> Optional[str]:
        normalized = str(value or "").strip()
        return normalized or None


class WatchlistScoreRefreshResult(BaseModel):
    symbol: str
    market: str
    status: str
    message: Optional[str] = None
    score: Optional[float] = None
    rank: Optional[int] = None
    scanner_run_id: Optional[int] = None


class WatchlistScoreRefreshResponse(BaseModel):
    ok: bool
    updated_count: int = 0
    failed_count: int = 0
    skipped_count: int = 0
    started_at: str
    completed_at: str
    markets: List[str] = Field(default_factory=list)
    results: List[WatchlistScoreRefreshResult] = Field(default_factory=list)


class WatchlistScoreRefreshStatusResponse(BaseModel):
    enabled: bool
    us_time: str
    cn_time: str
    hk_time: str
    max_symbols: int
    running: bool = False

File: apps/dsa-web/src/types/watchlist.ts
import type { InvestorSignalContract } from './scanner';

export interface WatchlistScannerLineageV1 {
  contractVersion: 'scanner_watchlist_lineage_v1' | string;
  source: 'scanner';
  scannerRunId?: number | null;
  symbol: string;
  market: string;
  rankAtScan?: number | null;
  scoreAtScan?: number | null;
  scoreSnapshotKind: 'saved_at_add' | 'post_add_refresh';
  runProfile?: string | null;
  runCompletedAt?: string | null;
  watchlistAddedAt?: string | null;
  themeId?: string | null;
  universeType?: string | null;
  researchReason: string;
  researchNextStep: string;
  dataState: 'available' | 'limited' | 'observation_only' | 'insufficient' | 'updating' | 'unavailable' | string;
  freshnessLabel: string;
  noAdviceBoundary: boolean;
  observationOnly: boolean;
  scoreGradeAllowed: boolean;
}

export interface WatchlistScannerIntelligence {
  lastScore?: number | null;
  lastRank?: number | null;
  status?: 'selected' | 'preview' | 'rejected' | 'data_failed' | 'unknown' | string | null;
  theme?: string | null;
  themeLabel?: string | null;
  profile?: string | null;
  reason?: string | null;
  lastScannedAt?: string | null;
  investorSignal?: InvestorSignalContract | null;
  scannerLineageV1?: WatchlistScannerLineageV1 | null;
}

export interface WatchlistScoreStatusContext {
  scope?: string | null;
  freshMeans?: string | null;
  sourceFreshnessImplied?: boolean | null;
  sourceAuthorityImplied?: boolean | null;
}

export interface WatchlistCatalystExposure {
  id: string;
  symbol: string;
  market: string;
  category: string;
  title: string;
  summary: string;
  evidenceStatus: string;
  evidenceLabels?: string[] | null;
  asOf?: string | null;
  publishedAt?: string | null;
  timeframe?: string | null;
  reasonCodes?: string[] | null;
  observationOnly?: boolean | null;
  sourceAuthorityAllowed?: boolean | null;
  scoreContributionAllowed?: boolean | null;
  decisionGrade?: boolean | null;
  calendarClaimAllowed?: boolean | null;
}

export interface WatchlistStrategySimulationIntelligence {
  lookbackDays?: number | null;
  forwardDays?: number | null;
  avgForwardReturnPct?: number | null;
  hitRate?: number | null;
  avgExcessReturnPct?: number | null;
  selectionCount?: number | null;
  dataCoverage?: number | null;
  status?: 'ready' | 'partial' | 'insufficient_history' | 'unknown' | string | null;
}

export interface WatchlistBacktestIntelligence {
  lastResultId?: number | null;
  totalReturnPct?: number | null;
  maxDrawdownPct?: number | null;
  sharpe?: number | null;
  tradeCount?: number | null;
  testedAt?: string | null;
}

export interface WatchlistIntelligence {
  scanner?: WatchlistScannerIntelligence | null;
  strategySimulation?: WatchlistStrategySimulationIntelligence | null;
  backtest?: WatchlistBacktestIntelligence | null;
  catalystExposures?: WatchlistCatalystExposure[] | null;
}

export interface WatchlistRowResearchIdentity {
  name?: string | null;
  exchange?: string | null;
  sector?: string | null;
  industry?: string | null;
  canonicalSymbol?: string | null;
  displaySymbol?: string | null;
  displayName?: string | null;
  identityState?: string | null;
}

export interface WatchlistRowResearchQuote {
  state: 'available' | 'missing' | 'stale' | 'unknown' | string;
  price?: number | null;
  changePercent?: number | null;
  asOf?: string | null;
}

export interface WatchlistRowScannerLineage {
  runId?: number | null;
  rank?: number | null;
  score?: number | null;
  status?: string | null;
  lastScoredAt?: string | null;
}

export interface WatchlistRowResearchPacketResponse {
  symbol: string;
  market: string;
  identity: WatchlistRowResearchIdentity;
  savedItemSource: string;
  quote: WatchlistRowResearchQuote;
  scannerLineage: WatchlistRowScannerLineage;
  researchStatus: 'ready' | 'partial' | 'blocked' | 'unknown' | string;
  researchReadiness?: WatchlistResearchReadiness | null;
  missingData: string[];
  nextDataAction: string;
  observationOnly: true;
  noAdviceDisclosure: string;
}

export interface WatchlistSymbolIdentity {
  canonicalSymbol: string;
  displaySymbol: string;
  market: string;
  exchange?: string | null;
  displayName?: string | null;
  identityState: string;
}

export interface WatchlistResearchReadiness {
  contractVersion?: string | null;
  state: 'available' | 'partial' | 'stale' | 'unavailable' | 'pending' | 'unknown' | string;
  freshnessState: 'available' | 'partial' | 'stale' | 'unavailable' | 'pending' | 'unknown' | string;
  identityState: string;
  lastReviewedAt?: string | null;
  scoreFreshnessImplied?: boolean | null;
  sourceAuthorityImplied?: boolean | null;
}

export interface WatchlistItem {
  id: number;
  symbol: string;
  market: string;
  identity?: WatchlistSymbolIdentity | null;
  name?: string | null;
  source: string;
  createdNew?: boolean | null;
  duplicateOfId?: number | null;
  scannerRunId?: number | null;
  scannerRank?: number | null;
  scannerScore?: number | null;
  lastScoredAt?: string | null;
  scoreSource?: string | null;
  scoreProfile?: string | null;
  scoreReason?: string | null;
  scoreStatus?: string | null;
  scoreStatusContext?: WatchlistScoreStatusContext | null;
  scoreError?: string | null;
  researchReadiness?: WatchlistResearchReadiness | null;
  themeId?: string | null;
  universeType?: string | null;
  notes?: string | null;
  intelligence?: WatchlistIntelligence | null;
  rowResearchPacket?: WatchlistRowResearchPacketResponse | null;
  createdAt?: string | null;
  updatedAt?: string | null;
}

export interface WatchlistItemListResponse {
  items: WatchlistItem[];
}

export interface WatchlistItemCreateRequest {
  symbol: string;
  market: string;
  name?: string;
  source?: 'scanner';
  scannerRunId?: number;
  scannerRank?: number;
  scannerScore?: number;
  themeId?: string | null;
  universeType?: string | null;
  notes?: string | null;
}

export interface WatchlistDeleteResponse {
  deleted: number;
}

export type WatchlistResearchPriorityTier = 'attention' | 'follow_up' | 'monitor';

export interface WatchlistResearchPriorityEvidenceAge {
  state: string;
  lastReviewedAt?: string | null;
}

export interface WatchlistResearchOverlayDrilldownTarget {
  label: string;
  route: string;
  section: string;
  reason: string;
}

export interface WatchlistResearchPriorityQueueItem {
  symbol: string;
  priorityTier: WatchlistResearchPriorityTier;
  priorityReasonSafeLabel: string;
  evidenceAge: WatchlistResearchPriorityEvidenceAge;
  missingEvidence: string[];
  suggestedResearchPath: WatchlistResearchOverlayDrilldownTarget[];
  observationOnly: true;
}

export interface WatchlistResearchOverlayResponse {
  schemaVersion: string;
  overlayState: string;
  researchSummary: string;
  researchPriorityQueue: WatchlistResearchPriorityQueueItem[];
  observationOnly: true;
  decisionGrade: false;
}

export interface WatchlistScoreRefreshRequest {
  market?: string;
  source?: 'scanner';
  theme?: string;
  symbols?: string[];
  force?: boolean;
}

export interface WatchlistScoreRefreshResult {
  symbol: string;
  market: string;
  status: string;
  message?: string | null;
  score?: number | null;
  rank?: number | null;
  scannerRunId?: number | null;
}

export interface WatchlistScoreRefreshResponse {
  ok: boolean;
  updatedCount: number;
  failedCount: number;
  skippedCount: number;
  startedAt: string;
  completedAt: string;
  markets: string[];
  results: WatchlistScoreRefreshResult[];
}

export interface WatchlistScoreRefreshStatus {
  enabled: boolean;
  usTime: string;
  cnTime: string;
  hkTime: string;
  maxSymbols: number;
  running: boolean;
}
